pil-analyzer/src/pil_analyzer.rs

```diff
@@ -144,8 +149,21 @@ impl PILAnalyzer {
 
         for PILFile(file) in files {
             self.current_namespace = Default::default();
-            for statement in file {
-                self.handle_statement(statement);
+            for ref statement in file {
+                if let PilStatement::TraitImplementation(sr, trait_impl) = statement {
```

Comment:
Why do you need to do it this early? This way we don't even see what is coming later on (we cannot guarantee that trait implementations are always after trait declarations).

I would have expected it to be right before type checking, or is there a reason to do it during statement processing?

Reply:
> Why do you need to do it this early? This way we don't even see what is coming later on (we cannot guarantee that trait implementations are always after trait declarations).

Since this step is just to collect the implementations, I thought this would be the right location. In fact, this is one of the few steps where we can go over all the statements and keep the implementations before they are 'lost'.

The only thing that bothers me is having to call StatementProcessor, to transform the parsed::Expression into analyzed::Expression, but I don't think I have another option.


> I would have expected it to be right before type checking, or is there a reason to do it during statement processing?

The rest of the unification happens before the type check.